SQL Database enhancements

l3ender opened the pull request:
##### SUMMARY
This PR updates the following:
* Bumps [sql SDK](https://pypi.org/project/azure-mgmt-sql/) to latest version.
* Adds following parameters to `azure_rm_sqlserver`:
  - `minimal_tls_version`: minimum TLS version allowed by the client.
  - `public_network_access`: enable/disable public network access.
  - `restrict_outbound_network_access`: enable/disable outbound network access.
  - `change_admin_password`: allow module usage for updating SQL password.
  - `administrators`: configure Azure Active Directory administrator for the server.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
* azure_rm_sqlserver
* azure_rm_sqlserver_info

##### ADDITIONAL INFORMATION

Using the following playbook at the root of this repo:

```yaml
---
- name: "Playbook for testing."
  hosts: "localhost"
  connection: "local"
  gather_facts: false
  vars:
    resource_group: "test-rg"
  collections:
    - azure.azcollection

  tasks:
    - name: "Include tests"
      include_tasks: tests/integration/targets/azure_rm_sqlserver/tasks/main.yml
```

l3ender committed:
prevent list index out of range errors when comparing dicts with empty lists

l3ender committed:
update sql sdk; support new fields and admin password change

l3ender commented:
@Fred-sun Here is another PR to review--thank you!

Fred-sun commented:
@l3ender Thank you for your contribution. We will push forward the import process of new SDK as soon as possible and review the PR. Thank you very much!

l3ender committed:
Merge branch 'dev' into sql-enhancements

l3ender committed:
Merge branch 'dev' into sql-enhancements

l3ender committed:
minor lint fixes

l3ender committed:
Merge remote-tracking branch 'upstream/dev' into sql-enhancements

l3ender committed:
disable log warning for non-sensitive parameter

l3ender committed:
correct sku idempotency check

l3ender committed:
Merge remote-tracking branch 'upstream/dev' into sql-enhancements

l3ender committed:
add support for azure ad admin on sql server

l3ender committed:
bump version added after module release

l3ender committed:
add tests for sql server with azure ad admin

l3ender committed:
add example for creating sql server with azure ad

l3ender committed:
add additional output for sql server + test coverage

l3ender committed:
Merge remote-tracking branch 'upstream/dev' into sql-enhancements

l3ender commented:
Hello @Fred-sun, can you please update on this PR? Thank you!!

Fred-sun commented:
@l3ender Thank you for your contribution, because our test account does not have permission to create adgroup-related resources, please delete the azure_rm_adgroup related use cases from the use case. Another, Please resolve the conflict file. Thank you!

Fred-sun commented on tests/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -67,6 +166,58 @@
+  assert:
+    that:
+      - output.changed == False
+      - output.servers['sqlsrv-azuread-{{ random_postfix }}'].id != None
```
In YAML, variable values should be referenced with "." and remove "''".  Change everything like this. Thank you very much!
```suggestion
      - output.servers.sqlsrv-azuread-{{ random_postfix }}.id != None
```

l3ender replied:
The output format of having `servers` be a list of dictionaries is existing functionality (see same in [master branch](https://github.com/l3ender/azure/blob/094fff8449a1e39deecc9945827f90c42cf7a8d7/plugins/modules/azure_rm_sqlserver_info.py#L51)).

Fred-sun commented on tests/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -47,6 +48,104 @@
+- name: Create extended instance of SQL Server
+  azure_rm_sqlserver:
+    resource_group: "{{ resource_group }}"
+    name: "sqlsrv-extended-{{ random_postfix }}"
```
The next step is to get the name and compare. Yaml treats "-" as a column when comparing, so it is recommended not to use "-" in the name. All of the following changes will be made. Thank you very much!
```suggestion
    name: "sqlsrvextended{{ random_postfix }}"
```

l3ender replied:
Can you please elaborate on why you don't want dashes in the name? It is valid YAML and there are existing test cases which do the same. Keeping the dashes improves visibility so I would prefer them to stay.

l3ender committed:
Merge remote-tracking branch 'upstream/dev' into sql-enhancements

l3ender committed:
disable test case as per review comment

l3ender commented:
Hi @Fred-sun, I have resolved the conflicts and updated the PR to default the `azure_rm_adgroup` related tasks to not execute. I would like to keep them as optional in case someone wants to run the whole test case from start to finish--this is helpful for someone like me to ensure no breaking changes occur when making updates.

Please take another look. Thank you!

Fred-sun commented:
@l3ender Thank you for your update. Do you have any test cases implemented? And you get a PASS. Thank you very much!

Fred-sun commented on tests/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -67,6 +168,58 @@
+  assert:
+    that:
+      - output.changed == False
+      - output.servers['sqlsrv-azuread-{{ random_postfix }}'].id != None
```
```suggestion
      - output.servers.sqlsrv-azuread-{{ random_postfix }}.id != None
```

l3ender replied:
The output format is existing (mentioned in https://github.com/ansible-collections/azure/pull/681#discussion_r823269611).

Fred-sun commented on tests/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -47,6 +49,105 @@
+      - output.changed
+
+- name: Create test Azure AD group
+  azure_rm_adgroup:
```
Remove the azure_rm_adgroup related use case that our automated project executes on a weekly basis. thank you

l3ender replied:
I've added a conditional to the `azure_rm_adgroup` related tasks so they are disabled by default (mentioned in https://github.com/ansible-collections/azure/pull/681#issuecomment-1063591029). Thank you!

l3ender committed:
correct test case which requires restore happen in same region

l3ender committed:
clean up sql tests related to aad group

l3ender commented:
Hi @Fred-sun! I have run the SQL tests using the following playbook. All tests have passed with no issue.

```yml
---
- name: "Playbook for testing."
  hosts: "localhost"
  connection: "local"
  gather_facts: false
  vars:
    resource_group: "automated-testing"
    resource_group_secondary: "automated-testing-secondary"
    azure_tenant: "5fa44015-7a98-419d-a210-311acc1216b9"
  collections:
    - azure.azcollection

  tasks:
    - name: "Include tests"
      include_tasks: "tests/integration/targets/azure_rm_sqlserver/tasks/main.yml"
```

I ran it like so:
```bash
ansible-playbook sandbox.yml -v
```

<details>
<summary>Click to view playbook output</summary>

```bash
(venv) ross@IT-LT207:/mnt/d/repos/azure$ ansible-playbook sandbox.yml -v
No config file found; using defaults
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match
'all'

PLAY [Playbook for testing.] *******************************************************************************************

TASK [Include tests] ***************************************************************************************************
included: /mnt/d/repos/azure/tests/integration/targets/azure_rm_sqlserver/tasks/main.yml for localhost

TASK [Prepare random number] *******************************************************************************************
ok: [localhost] => {"ansible_facts": {"random_postfix": "969ef45326", "sql_admin_group": "Test Security Group", "tenant_id": "5fa44015-7a98-419d-a210-311acc1216b9"}, "changed": false}

TASK [Create instance of SQL Server -- check mode] *********************************************************************
[WARNING]: Azure API profile latest does not define an entry for SqlManagementClient
changed: [localhost] => {"changed": true}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of SQL Server] ***********************************************************************************
changed: [localhost] => {"changed": true, "fully_qualified_domain_name": "sqlsrv969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326", "state": "Ready", "version": "12.0"}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create again instance of SQL Server] *****************************************************************************
ok: [localhost] => {"changed": false, "compare": ["new item was empty, using old [][ tags ]"], "fully_qualified_domain_name": "sqlsrv969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326", "modifiers": {}, "state": "Ready", "version": "12.0"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create extended instance of SQL Server] **************************************************************************
changed: [localhost] => {"changed": true, "fully_qualified_domain_name": "sqlsrv-extended-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-extended-969ef45326", "state": "Ready", "version": "12.0"}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create extended instance of SQL Server - idempotent] *************************************************************
ok: [localhost] => {"changed": false, "compare": ["new item was empty, using old [][ tags ]"], "fully_qualified_domain_name": "sqlsrv-extended-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-extended-969ef45326", "modifiers": {}, "state": "Ready", "version": "12.0"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Update SQL admin password] ***************************************************************************************
changed: [localhost] => {"changed": true, "compare": ["changed [/administrator_login_password] ******** != None - default"], "fully_qualified_domain_name": "sqlsrv-extended-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-extended-969ef45326", "modifiers": {}, "state": "Ready", "version": "12.0"}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Get Azure AD group info] *****************************************************************************************
ok: [localhost] => {"ad_groups": [{"display_name": "Test Security Group", "mail": null, "mail_enabled": false, "mail_nickname": "3fe5305c-0", "object_id": "2801231e-a95d-4f7f-a768-133a4a14f8a7", "security_enabled": true}], "changed": false}

TASK [Assert AD group exists] ******************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of SQL Server with Azure AD admin] ***************************************************************
changed: [localhost] => {"changed": true, "fully_qualified_domain_name": "sqlsrv-azuread-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-azuread-969ef45326", "state": "Ready", "version": "12.0"}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of SQL Server with Azure AD admin - idempotent] **************************************************
ok: [localhost] => {"changed": false, "compare": [], "fully_qualified_domain_name": "sqlsrv-azuread-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-azuread-969ef45326", "modifiers": {}, "state": "Ready", "version": "12.0"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Server] *****************************************************************************************
ok: [localhost] => {"changed": false, "servers": {"sqlsrv969ef45326": {"admin_username": "mylogin", "administrators": null, "fully_qualified_domain_name": "sqlsrv969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326", "kind": "v12.0", "location": "eastus", "minimal_tls_version": null, "name": "sqlsrv969ef45326", "public_network_access": "Enabled", "restrict_outbound_network_access": "Disabled", "state": "Ready", "tags": {"aaa": "bbb"}, "type": "Microsoft.Sql/servers", "version": "12.0"}}}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts for extended SQL Server] ****************************************************************************
ok: [localhost] => {"changed": false, "servers": {"sqlsrv-extended-969ef45326": {"admin_username": "mylogin2", "administrators": null, "fully_qualified_domain_name": "sqlsrv-extended-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-extended-969ef45326", "kind": "v12.0", "location": "eastus", "minimal_tls_version": "1.2", "name": "sqlsrv-extended-969ef45326", "public_network_access": "Disabled", "restrict_outbound_network_access": "Enabled", "state": "Ready", "tags": {"aaa": "bbb"}, "type": "Microsoft.Sql/servers", "version": "12.0"}}}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts for SQL Server with Azure AD admin] *****************************************************************
ok: [localhost] => {"changed": false, "servers": {"sqlsrv-azuread-969ef45326": {"admin_username": "sqllogin", "administrators": {"administrator_type": "ActiveDirectory", "azure_ad_only_authentication": false, "login": "Test Security Group", "principal_type": "Group", "sid": "2801231e-a95d-4f7f-a768-133a4a14f8a7", "tenant_id": "5fa44015-7a98-419d-a210-311acc1216b9"}, "fully_qualified_domain_name": "sqlsrv-azuread-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-azuread-969ef45326", "kind": "v12.0", "location": "eastus", "minimal_tls_version": null, "name": "sqlsrv-azuread-969ef45326", "public_network_access": "Enabled", "restrict_outbound_network_access": "Disabled", "state": "Ready", "tags": {}, "type": "Microsoft.Sql/servers", "version": "12.0"}}}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Server - unexisting] ****************************************************************************
ok: [localhost] => {"changed": false, "servers": {}}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Server - list] **********************************************************************************
ok: [localhost] => {"changed": false, "servers": {"sqlsrv-azuread-969ef45326": {"admin_username": "sqllogin", "administrators": {"administrator_type": "ActiveDirectory", "azure_ad_only_authentication": false, "login": "Test Security Group", "principal_type": "Group", "sid": "2801231e-a95d-4f7f-a768-133a4a14f8a7", "tenant_id": "5fa44015-7a98-419d-a210-311acc1216b9"}, "fully_qualified_domain_name": "sqlsrv-azuread-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-azuread-969ef45326", "kind": "v12.0", "location": "eastus", "minimal_tls_version": null, "name": "sqlsrv-azuread-969ef45326", "public_network_access": "Enabled", "restrict_outbound_network_access": "Disabled", "state": "Ready", "tags": {}, "type": "Microsoft.Sql/servers", "version": "12.0"}, "sqlsrv-extended-969ef45326": {"admin_username": "mylogin2", "administrators": null, "fully_qualified_domain_name": "sqlsrv-extended-969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv-extended-969ef45326", "kind": "v12.0", "location": "eastus", "minimal_tls_version": "1.2", "name": "sqlsrv-extended-969ef45326", "public_network_access": "Disabled", "restrict_outbound_network_access": "Enabled", "state": "Ready", "tags": {"aaa": "bbb"}, "type": "Microsoft.Sql/servers", "version": "12.0"}, "sqlsrv969ef45326": {"admin_username": "mylogin", "administrators": null, "fully_qualified_domain_name": "sqlsrv969ef45326.database.windows.net", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326", "kind": "v12.0", "location": "eastus", "minimal_tls_version": null, "name": "sqlsrv969ef45326", "public_network_access": "Enabled", "restrict_outbound_network_access": "Disabled", "state": "Ready", "tags": {"aaa": "bbb"}, "type": "Microsoft.Sql/servers", "version": "12.0"}}}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of SQL Database -- check mode] *******************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of SQL Database] *********************************************************************************
changed: [localhost] => {"changed": true, "database_id": "6ba2eb43-2c9a-40f4-9dc2-a72150fe3a03", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326", "status": "Online"}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create again instance of SQL Database] ***************************************************************************
ok: [localhost] => {"changed": false, "database_id": "6ba2eb43-2c9a-40f4-9dc2-a72150fe3a03", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326", "status": "Online"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Database and wait for restore point] ************************************************************
FAILED - RETRYING: [localhost]: Gather facts SQL Database and wait for restore point (10 retries left).
FAILED - RETRYING: [localhost]: Gather facts SQL Database and wait for restore point (9 retries left).
ok: [localhost] => {"attempts": 3, "changed": false, "databases": [{"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": "2022-03-10T18:41:21.000Z", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326", "kind": "v12.0,user", "location": "eastus", "name": "database969ef45326", "resource_group": "automated-testing", "sku": {"capacity": 250, "name": "P2", "tier": "Premium"}, "status": "Online", "tags": {"aaa": "bbb"}, "zone_redundant": false}]}

TASK [Assert that it can be restored from] *****************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create second SQL Database, restoring from the previous Database] ************************************************
changed: [localhost] => {"changed": true, "database_id": "daec2188-7374-4c73-878c-f80d002b6695", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326PITR", "status": "Online"}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of SQL Database Point in time recovery] **********************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create second SQL Database] **************************************************************************************
changed: [localhost] => {"changed": true, "database_id": "e422ee4e-2137-4b22-b045-0dca32ee071e", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326second", "status": "Online"}

TASK [Gather facts SQL Database] ***************************************************************************************
ok: [localhost] => {"changed": false, "databases": [{"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": "2022-03-10T18:41:21.000Z", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326", "kind": "v12.0,user", "location": "eastus", "name": "database969ef45326", "resource_group": "automated-testing", "sku": {"capacity": 250, "name": "P2", "tier": "Premium"}, "status": "Online", "tags": {"aaa": "bbb"}, "zone_redundant": false}]}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Database] ***************************************************************************************
ok: [localhost] => {"changed": false, "databases": [{"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": null, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326second", "kind": "v12.0,user,vcore", "location": "eastus", "name": "database969ef45326second", "resource_group": "automated-testing", "sku": {"capacity": 2, "name": "GP_Gen5_2", "tier": "GeneralPurpose"}, "status": "Online", "tags": null, "zone_redundant": false}, {"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": null, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/master", "kind": "v12.0,system", "location": "eastus", "name": "master", "resource_group": "automated-testing", "sku": {"capacity": 0, "name": "System2", "tier": "System"}, "status": "Online", "tags": null, "zone_redundant": false}, {"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": "2022-03-10T18:41:21.000Z", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef45326", "kind": "v12.0,user", "location": "eastus", "name": "database969ef45326", "resource_group": "automated-testing", "sku": {"capacity": 250, "name": "P2", "tier": "Premium"}, "status": "Online", "tags": {"aaa": "bbb"}, "zone_redundant": false}]}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of secondary database] ***************************************************************************
changed: [localhost] => {"changed": true}

TASK [Delete instance of SQL Database -- check mode] *******************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of SQL Database] *********************************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete unexisting instance of SQL Database] **********************************************************************
ok: [localhost] => {"changed": false}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create SQL Database with sku -- check mode] **********************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create SQL Database with sku] ************************************************************************************
changed: [localhost] => {"changed": true, "database_id": "7c6d7980-0310-46ab-b62c-9e84428c4401", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef453262", "status": "Online"}

TASK [Assert the resource instance is well created with good SKU] ******************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Database with good SKU] *************************************************************************
ok: [localhost] => {"changed": false, "databases": [{"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": null, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef453262", "kind": "v12.0,user", "location": "eastus", "name": "database969ef453262", "resource_group": "automated-testing", "sku": {"capacity": 10, "name": "S0", "tier": "Standard"}, "status": "Online", "tags": {"aaa": "bbb"}, "zone_redundant": false}]}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create again instance of SQL Database with same SKU] *************************************************************
ok: [localhost] => {"changed": false, "database_id": "7c6d7980-0310-46ab-b62c-9e84428c4401", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef453262", "status": "Online"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create again instance of SQL Database with New SKU] **************************************************************
changed: [localhost] => {"changed": true, "database_id": "7c6d7980-0310-46ab-b62c-9e84428c4401", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef453262", "status": "Online"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Database with good New SKU] *********************************************************************
ok: [localhost] => {"changed": false, "databases": [{"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": null, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/database969ef453262", "kind": "v12.0,user", "location": "eastus", "name": "database969ef453262", "resource_group": "automated-testing", "sku": {"capacity": 125, "name": "P1", "tier": "Premium"}, "status": "Online", "tags": {"aaa": "bbb"}, "zone_redundant": false}]}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of SQL Database] *********************************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Database] ***************************************************************************************
ok: [localhost] => {"changed": false, "databases": []}

TASK [Assert that empty dictionary was returned] ***********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Database] ***************************************************************************************
ok: [localhost] => {"changed": false, "databases": [{"collation": "SQL_Latin1_General_CP1_CI_AS", "earliest_restore_date": null, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/databases/master", "kind": "v12.0,system", "location": "eastus", "name": "master", "resource_group": "automated-testing", "sku": {"capacity": 0, "name": "System2", "tier": "System"}, "status": "Online", "tags": null, "zone_redundant": false}]}

TASK [Assert that empty dictionary was returned (one database is there by default)] ************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of Firewall Rule -- check mode] ******************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create instance of Firewall Rule] ********************************************************************************
changed: [localhost] => {"changed": true, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/firewallRules/firewallrule969ef45326"}

TASK [Assert the resource instance is well created] ********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create again instance of Firewall Rule] **************************************************************************
ok: [localhost] => {"changed": false, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/firewallRules/firewallrule969ef45326"}

TASK [Assert the state has not changed] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Create Firewall Rule - second] ***********************************************************************************
changed: [localhost] => {"changed": true, "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/firewallRules/firewallrule969ef45326second"}

TASK [Gather facts SQL Firewall Rule] **********************************************************************************
ok: [localhost] => {"changed": false, "rules": [{"end_ip_address": "172.28.10.138", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/firewallRules/firewallrule969ef45326", "name": "firewallrule969ef45326", "resource_group": "automated-testing", "server_name": "sqlsrv969ef45326", "start_ip_address": "172.28.10.136"}]}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Gather facts SQL Firewall Rule] **********************************************************************************
ok: [localhost] => {"changed": false, "rules": [{"end_ip_address": "172.28.10.138", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/firewallRules/firewallrule969ef45326", "name": "firewallrule969ef45326", "resource_group": "automated-testing", "server_name": "sqlsrv969ef45326", "start_ip_address": "172.28.10.136"}, {"end_ip_address": "172.28.10.138", "id": "/subscriptions/15d7b290-260a-4ea7-a138-37b3f781f644/resourceGroups/automated-testing/providers/Microsoft.Sql/servers/sqlsrv969ef45326/firewallRules/firewallrule969ef45326second", "name": "firewallrule969ef45326second", "resource_group": "automated-testing", "server_name": "sqlsrv969ef45326", "start_ip_address": "172.28.10.136"}]}

TASK [Assert that facts are returned] **********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of Firewall Rule] ********************************************************************************
changed: [localhost] => {"changed": true}

TASK [Delete instance of Firewall Rule] ********************************************************************************
changed: [localhost] => {"changed": true}

TASK [Gather facts SQL Firewall Rule] **********************************************************************************
ok: [localhost] => {"changed": false, "rules": []}

TASK [Assert that empty dictionary was returned] ***********************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of SQL Server -- check mode] *********************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete instance of SQL Server] ***********************************************************************************
changed: [localhost] => {"changed": true}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete unexisting instance of SQL Server] ************************************************************************
ok: [localhost] => {"changed": false}

TASK [Assert the state has changed] ************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [Delete extended instance of SQL Server] **************************************************************************
changed: [localhost] => {"changed": true}

TASK [Delete instance of SQL Server with Azure AD admin] ***************************************************************
changed: [localhost] => {"changed": true}

PLAY RECAP *************************************************************************************************************
localhost                  : ok=95   changed=26   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```
</details>

Let me know if you have any other questions. Thank you!

l3ender commented:
@Fred-sun Were any other changes needed for this PR?

l3ender committed:
Merge branch 'dev' into sql-enhancements

Fred-sun commented on tests/integration/targets/azure_rm_sqlserver/tasks/main.yml:
```diff
@@ -47,6 +49,108 @@
+    that:
+      - output.changed
+
+- name: Get Azure AD group info
```
Because this use case uses our pipeline environment but does not have permissions to operate azure_rm_adgroup *. You are advised to delete it. Thank you very much!

l3ender replied:
My apologies, @Fred-sun! I thought that your pipeline just restricted creation of groups and not also retrieving them.

I have updated to remove azure_rm_adgroup* tasks. Please check again. Thank you!

l3ender committed:
remove tasks loading ad group info

Fred-sun commented:
@l3ender 

```e,g,

l3ender committed:
conditionally run azuread tasks

l3ender commented:
Thanks @Fred-sun! I have modified the playbook so that AzureAD tasks will not run unless explicitly specified at time of run. Please take a look--thank you!
